[DBZ-3417] fix on-the-fly schema changes

smiklosovic opened the pull request:
Hi @gunnarmorling / @bingqinzhou 

we wanted to use your connector (and we still do) but we found couple critical issues with the current solution you are providing which renders this connector not fully suitable to production needs.

The problem we are seeing is quite hard to grasp and fully explain so I prepared two "readmes" here

part 1: this is our original attempt to solve it (1)
part 2: this is rewritten part 1 because we identified that the solution in part 1 is not enough (2)

We are providing the "part 2" solution in this PR.

We know this is quite big change but without it we are very hesitant to use it and our customers would find it very hard to use - if not impossible. We have our fix internally but we want this change to happen in upstream too for a greater good as well as to not balkanize this space with custom modifications.

It would be very nice if you could give us a constructive feedback on what we did and eventually we might cooperate to have this merged in your main branch. I am here to fully cooperate.

Please keep in mind that this serves as a basic building block to move this forward and open more deeper discussion about it.

1) https://github.com/debezium/debezium-connector-cassandra/blob/ac43b7797c084c3e67cedde3662af1e58de8a4c2/REPORT.adoc
2) https://github.com/debezium/debezium-connector-cassandra/blob/ac43b7797c084c3e67cedde3662af1e58de8a4c2/REPORT_2.adoc

Regards, Stefan

EDIT:

this has DDD-2 where reports were consolidated https://github.com/debezium/debezium-design-documents/pull/2

jpechane commented:
@smiklosovic Hi, thanks for the report! Could you please turn this into a regular PR? Please raise a Jira with the report findings - you know the drill :-). Depending on the scope of change a design docuemnt can be raise here too https://github.com/debezium/debezium-design-documents

smiklosovic commented:
@jpechane I was not able to create a ticket yet, it seems like Jira hangs ... I ll keep trying

smiklosovic commented:
@jpechane I created by mistake DBZ-341{8,9} and 3420 as I was thinking that creation of a ticket just hangs but it created it but I was not redirected so I kept trying. I try to close remaining ones (3417 being the original one I will use) but the cancellation hangs too ... Feel free to close the other two if you manage to do that.

smiklosovic commented:
@jpechane https://github.com/debezium/debezium-design-documents/pull/2

bingqinzhou commented:
@smiklosovic Thanks for the report! I can reproduce the issue, looking at it now.

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,202 @@
+                logger.info("changed table {}", newTableMetadata.asCQLQuery());
+            }
+        };
+    }
```
I'm concerned about the complicated logic and codes involved in these handler functions above for manually updating Cassandra schema.instance or metadata:
- Is there any wrapped-up function in Cassandra library that can be used here directly?
- How much test have you done to ensure the correctness of the manual codes so far?
- Any expensive operation involved in these functions? Like: 
```
 - Keyspace.open(newCFMetaData.ksName).initCf(newCFMetaData, false);
 - Keyspace.openWithoutSSTables(ksName)
```

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,202 @@
+
+    @Override
+    public void initialize() {
+        final Map<KeyspaceTable, SchemaHolder.KeyValueSchema> tables = getTableMetadata()
```
This should use getCdcEnabledTableMetadataList() in SchemaHolder here, as Cassandra Connector should only care about and hold schemas for CDC Enabled tables.

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,202 @@
+        schemaHolder.tableToKVSchemaMap.clear();
+    }
+
+    private List<TableMetadata> getTableMetadata() {
```
This function shouldn't be needed.

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,202 @@
+            @Override
+            public void onTableAdded(final TableMetadata table) {
+                logger.info(String.format("Table %s.%s detected to be added!", table.getKeyspace().getName(), table.getName()));
+                schemaHolder.addTable(new KeyspaceTable(table),
```
Only `CDC enabled` tables should be added to schemaHolder.

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,202 @@
+            @Override
+            public void onTableRemoved(final TableMetadata table) {
+                logger.info(String.format("Table %s.%s detected to be removed!", table.getKeyspace().getName(), table.getName()));
+                schemaHolder.removeTable(new KeyspaceTable(table));
```
Only `CDC enabled` tables should be removed from schemaHolder.

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,202 @@
+                        oldTableMetaData.getOptions().isCDC(),
+                        newTableMetadata.getOptions().isCDC()));
+
+                schemaHolder.addTable(new KeyspaceTable(newTableMetadata),
```
Only `CDC enabled` tables should be added/updated in schemaHolder.

bingqinzhou commented:
@smiklosovic I left comments in your PR and thank you for digging into this issue! 
To determine if we're going to use this approach in Cassandra Connector, I want to confirm about the following questions:
- What is the expected delay for `SchemaChangeListener` ? For example, if a table metadata is changed, does `onTableChanged` gets triggered immediately or after some specific amount of time?
- How do you guarantee the correctness of the codes for manually updating schema.instance and metadata so far in the listener?
- Is there any expensive operations involved in the codes for manually updating schema.instance and metadata?

smiklosovic commented:
Hi @bingqinzhou ,

Check this in a driver:

````
            this.schemaRefreshRequestDebouncer = new EventDebouncer<SchemaRefreshRequest>(
                    "Schema refresh",
                    scheduledTasksExecutor,
                    new SchemaRefreshRequestDeliveryCallback()
            ) {

                @Override
                int maxPendingEvents() {
                    // defaults to 20
                    return configuration.getQueryOptions().getMaxPendingRefreshSchemaRequests();
                }

                @Override
                long delayMs() {
                    // defaults to 1000 (1s)
                    return configuration.getQueryOptions().getRefreshSchemaIntervalMillis();
                }
            };
````

Javadoc on that debouncer is:

````
/**
 * A helper class to debounce events received by the Control Connection.
 * <p/>
 * This class accumulates received events, and delivers them when either:
 * - no events have been received for delayMs
 * - maxPendingEvents have been received
 */
````

* I can write a test for this, it would basically cover the scenario we are seeing here. Apart from that I am not sure what "to guarantee" means. How would you like me approaching this then? 
* I do not think so. This is what Cassandra is doing internally and I bet they spent a great deal of time to have it as performance friendly as possible. To add to it, we are not touching anything at disk.

Now about the "only cdc tables should be tracked" comments you wrote across this PR. This is not so simple as it seems.

A commit log consists of mutations. Mutations consists of PartitionUpdates. If a log is copied to CDC directory, obviously, there are also Mutations in that log segment which are *not* tracked by cdc but commit log handler will read them anyway.

Only your check by "trackedByCDC" in handler's method will "filter them".

But in order to pass that Mutation into handler, it is already deserialised.

In order to deserialise it, this code is called (1) which in turn calls this (2). This (2) is crucial - it deserialised UUID and it goes to fetch CFMetaData - *but that table does not have to be there* because schema is *not* updated _in Cassandra_. The fact that you are caching only "cdc-enabled" does not matter in the broader sense because regardless if you do or you do not, there are going to be be Mutations deserialised which are not cdc-enabled (and you filter them afterwards) but they will be deserialised anyway.

So if we do not track tables which are not-cdc enabled, that deserialisation method might return null and it would throw etc etc ... Hence we need to track everything. 

(1) https://github.com/apache/cassandra/blob/181a4969290f1c756089b2993a638fe403bc1314/src/java/org/apache/cassandra/db/partitions/PartitionUpdate.java#L913

(2) https://github.com/apache/cassandra/blob/181a4969290f1c756089b2993a638fe403bc1314/src/java/org/apache/cassandra/config/CFMetaData.java#L1571-L1572

smiklosovic commented on src/main/java/io/debezium/connector/cassandra/NoOpSchemaChangeListener.java:
```diff
@@ -0,0 +1,117 @@
+import com.datastax.driver.core.TableMetadata;
+import com.datastax.driver.core.UserType;
+
+public class NoOpSchemaChangeListener implements SchemaChangeListener {
```
answered below this issue

smiklosovic commented on src/main/java/io/debezium/connector/cassandra/CommitLogReadHandlerImpl.java:
```diff
@@ -303,18 +303,22 @@ private void process(PartitionUpdate pu, OffsetPosition offsetPosition, Keyspace
 
-        SchemaHolder.KeyValueSchema keyValueSchema = schemaHolder.getOrUpdateKeyValueSchema(keyspaceTable);
+        SchemaHolder.KeyValueSchema keyValueSchema = schemaHolder.getKeyValueSchema(keyspaceTable);
+        if (keyValueSchema == null) {
```
Imagine there is a bunch of commit logs in cdc and your processor / handler are reading them (Mutations) but at the same time somebody dropped that table via cql. So what happens is that a listener is invoked and it would be removed from the holder but you might still read a Mutation which was just on cdc before but it does not exist anymore and holder does not hold it either, but you are still trying to send it to Kafka.

bingqinzhou replied:
Nice catch! Can you also add a log here please? Something like:
```
LOGGER.warn("Unable to get KeyValueSchema for table {}. It might have been deleted or CDC disabled.", keyspaceTable.toString()).
```

smiklosovic commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -70,10 +60,36 @@ public KeyValueSchema getOrUpdateKeyValueSchema(KeyspaceTable kt) {
                 .collect(Collectors.toSet());
     }
 
+    public synchronized void removeKeyspace(String keyspace) {
```
when somebody `DROP KEYSPACE abc`, listener will be invoked on `onKeyspaceRemoved` and keyspaces will be removed so we need to reflect this in schema holder too, what is the purpose of holding some schemas for tables belonging to a keyspace which does not exist anymore?

bingqinzhou replied:
Can you rename this function to be something like `removeSchemasOfAllTablesInKeyspace()` please?

smiklosovic commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -70,10 +60,36 @@ public KeyValueSchema getOrUpdateKeyValueSchema(KeyspaceTable kt) {
+        keyspaceTableMap.put(kst.keyspace, kst.table);
+    }
+
+    public synchronized boolean contains(String keyspace, String table) {
```
nowhere, had to be some leftover, deleted.

smiklosovic commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,202 @@
+
+        schemaChangeListener = new NoOpSchemaChangeListener() {
+            @Override
+            public void onKeyspaceAdded(final KeyspaceMetadata keyspace) {
```
well that might be true in general but without having a keyspace added, the following update / creation of a table will not be possible because the underlying keyspace will not be there as we would just ignore its creation as you propose. The keyspace has to be there for the next tables you are going to possibly add afterwards.

smiklosovic commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,202 @@
+            }
+
+            @Override
+            public void onKeyspaceRemoved(final KeyspaceMetadata keyspace) {
```
that is not the case

https://github.com/datastax/java-driver/blob/3.5.0/driver-core/src/main/java/com/datastax/driver/core/DefaultResultSetFuture.java#L82-L114

removing a keyspace does not trigger removing a table methods

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -31,49 +37,64 @@
+    }
+
+    public synchronized void removeTableSchema(KeyspaceTable kst) {
+        tableToKVSchemaMap.remove(kst);
```
Can you check if `kst` exists in map before trying to remove it please?

smiklosovic replied:
it is not necessary, IDEA even highlights this

bingqinzhou replied:
> it is not necessary, IDEA even highlights this

It might miss something but I still don't understand why this is not necessary. I'm confused about what `IDEA even highlights this` means. Can you explain more clearly?

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaHolder.java:
```diff
@@ -31,49 +37,64 @@
 
-    private final Map<KeyspaceTable, KeyValueSchema> tableToKVSchemaMap = new ConcurrentHashMap<>();
+    public final Map<KeyspaceTable, KeyValueSchema> tableToKVSchemaMap = new ConcurrentHashMap<>();
+    public final Multimap<String, String> keyspaceTableMap = Multimaps.synchronizedListMultimap(ArrayListMultimap.create());
```
What is this `keyspaceTableMap` used for here?

bingqinzhou replied:
Hi can you explain what this `keyspaceTableMap` is used for?

smiklosovic replied:
It was a remaining of the old solution where I was holding some other map paralel to this one because if I wanted to ask if some key is there, I had to construct a new object every time. If you ask if some key is present in tableToKVSchemaMap, you have to do something like 

````
if (tableToKVSchemaMap.contains(new KeyspaceTable(...)) {
   ...
}
````

and creating an object here is just no too much performance friendly if it is going to be asked frequently. Having a paralel structure based just on a <String,String> is way way easier.

But I do not think we need that now, do we ... it is not used anywhere (but it was previously and I just forgot to remove it).

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,216 @@
+
+                LOGGER.info("Updated schema for table {}", newTableMetadata.asCQLQuery());
+            }
+        };
```
As it's difficult to write tests for this listener from your point, can you wrap the content in each function in this listener into a try-catch block and log error when it fails please? Like:

```
onTableChanged() {
       try {
              // update schemas 
       catch () {
             // LOGGER.error()
       }
 }
```

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/SchemaProcessor.java:
```diff
@@ -5,24 +5,216 @@
+                    // if new table is not on cdc anymore, remove it
+                    schemaHolder.removeTableSchema(new KeyspaceTable(newTableMetadata));
+                }
+                else if (newTableMetadata.getOptions().isCDC() && oldTableMetaData.getOptions().isCDC()) {
```
This can be combined with `if (newTableMetadata.getOptions().isCDC() && !oldTableMetaData.getOptions().isCDC())`.

smiklosovic replied:
i prefer to leave it this way, it reads better in my mind instead of having these ifs embedded

bingqinzhou replied:
> i prefer to leave it this way, it reads better in my mind instead of having these ifs embedded

I prefer these two conditions to be combined. It might seem clearer for the person who wrote the codes, but it's confusing and redundant for other people.

bingqinzhou commented on src/main/java/io/debezium/connector/cassandra/CommitLogReadHandlerImpl.java:
```diff
@@ -362,14 +366,17 @@ private void handlePartitionDeletion(PartitionUpdate pu, OffsetPosition offsetPo
-        Schema valueSchema = schema.valueSchema();
+        SchemaHolder.KeyValueSchema keyValueSchema = schemaHolder.getKeyValueSchema(keyspaceTable);
+        if (keyValueSchema == null) {
+            return;
```
Ditto, can you also add a log here please?

gunnarmorling commented:
Hey all, I'm in the process of catching up with this. I'm wondering: would it make sense to have a call with everyone, so we can talk about the problem and suggested solution? Details are still blurry to me, but this might be just me as I'm not too familiar with this connector.

smiklosovic commented:
@bingqinzhou added yet another commit with changes, please review

smiklosovic commented:
@bingqinzhou I am getting error on build, not sure what is wrong with that ... something on your side - Java version or similar.

smiklosovic commented:
> Hey all, I'm in the process of catching up with this. I'm wondering: would it make sense to have a call with everyone, so we can talk about the problem and suggested solution? Details are still blurry to me, but this might be just me as I'm not too familiar with this connector.

Hi @gunnarmorling , I do not have any problem with talking to you as such but I would very appreciate if you went through the DDD-2 and respective reports I put together very elaboratively just for this purpose to give you all information. I think that having a call would be just me reiterating what I wrote and it would be honestly just a waste of everybody's time. Feel free to raise any questions you have if you still do not understand it.

bingqinzhou commented:
> @bingqinzhou I am getting error on build, not sure what is wrong with that ... something on your side - Java version or similar.

The build failure seems to be related to a recent java 11 change. @jpechane or @gunnarmorling should have context about this.

bingqinzhou commented:
> Hey all, I'm in the process of catching up with this. I'm wondering: would it make sense to have a call with everyone, so we can talk about the problem and suggested solution? Details are still blurry to me, but this might be just me as I'm not too familiar with this connector.

Hi Gunnar, the cause and solution is explained in these two docs:
1. https://github.com/debezium/debezium-connector-cassandra/blob/ac43b7797c084c3e67cedde3662af1e58de8a4c2/REPORT.adoc
2. https://github.com/debezium/debezium-connector-cassandra/blob/ac43b7797c084c3e67cedde3662af1e58de8a4c2/REPORT_2.adoc illustrated in these docs. 

It also took me some time to read and understand the context, but the docs above clearly explains the issue and solution. The solution looks correct to me and I think it fixes the bug mentioned here as well as a know restriction in Cassandra Connector.

If you feel needed, I'd like to join the call for sure. My working time is 10AM-6PM PST. Please let me know if there's any workable time slot among all.

bingqinzhou commented:
@smiklosovic The code looks good to me. Can you rebase the PR to fix the build issue and manually test against the PR to make sure it works as expected with all these changes please? Thanks!

smiklosovic commented:
@bingqinzhou I rebased and squashed, I need your approval here as when I contribute for the first time I need approval to run a workflow (hence build).

smiklosovic commented:
@bingqinzhou still same error? am I doing something wrong here?

bingqinzhou commented:
> @bingqinzhou still same error? am I doing something wrong here?

@smiklosovic Maybe the rebase wasn't successful? Can you double check if your branch contains the change in this PR https://github.com/debezium/debezium-connector-cassandra/pull/27/files?

bingqinzhou commented:
@jpechane @gunnarmorling The build is failing for some java-version enforcement issue which seems to be related to the change in https://github.com/debezium/debezium-connector-cassandra/pull/27. Could you help take a look please?

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M2:enforce (enforce-java-version) on project debezium-build-parent: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error: Process completed with exit code 1.
```

smiklosovic commented:
yes, it is there https://i.postimg.cc/G2bVXpps/Screenshot-from-2021-04-28-08-55-13.png

ahmedjami commented:
Hi @gunnarmorling :)

could you please take a look at this PR when you get a chance? So, we can test it in our environment :)
Thanks in advance.

gunnarmorling commented:
Yes, will do. It may take me a bit though, I haven't really looked into this matter yet. In turn, any comment here will be appreciated: https://groups.google.com/g/debezium/c/aQjzdTcv2LA. Thanks :)

smiklosovic commented:
I am still not able to have a clean build. I think this issue is still on your side.

bingqinzhou commented:
> I am still not able to have a clean build. I think this issue is still on your side.

@smiklosovic Please update the commit message for f144c3c to begin with JIRA number. Currently re-build is blocked.

smiklosovic commented:
@bingqinzhou  better? I might squash it afterwards ...

bingqinzhou commented:
It seems that the build is still failing because of Java 11 enforcement:

```
Warning:  Rule 1: org.apache.maven.plugins.enforcer.RequireJavaVersion failed with message:
To build this project JDK 11 (or greater) is required. Please install it.
```
```
Error:  Failed to execute goal org.apache.maven.plugins:maven-enforcer-plugin:3.0.0-M2:enforce (enforce-java-version) on project debezium-build-parent: Some Enforcer rules have failed. Look above for specific messages explaining why the rule failed. -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
Error: Process completed with exit code 1.
```

smiklosovic commented:
@bingqinzhou would you mind to approve? it seems like something build-related just landed in `main`

smiklosovic commented:
Yay!

smiklosovic commented:
Please give me one more day here, I ll do final testing on that.

ahmedjami commented:
Great ! :)

smiklosovic commented:
Hey, I tested it, it works ... you may proceed to merge.

Thank you!

bingqinzhou commented:
> Hey, I tested it, it works ... you may proceed to merge.
> 
> Thank you!

Merged, thank you!

bingqinzhou commented:
Hi @smiklosovic, I'm seeing the following error during testing, which seems to be a bug with `onTableRemoved()` in `SchemaProcessor`. Can you please take a look when you get a chance, thanks!

```
2021-05-13 23:29:19,711 ERROR  ||  Error happened while removing table CREATE KEYSPACE test_keyspace WITH REPLICATION = { 'class' : 'org.apache.cassandra.locator.SimpleStrategy', 'replication_factor': '1' } AND DURABLE_WRITES = false;.table2  
java.lang.IllegalArgumentException: Unknown keyspace/cf pair (test_keyspace.table2)
	at org.apache.cassandra.db.Keyspace.getColumnFamilyStore(Keyspace.java:198)
	at io.debezium.connector.cassandra.SchemaProcessor$1.onTableRemoved(SchemaProcessor.java:149)
	at com.datastax.driver.core.Metadata.triggerOnTableRemoved(Metadata.java:662)
	at com.datastax.driver.core.Cluster$Manager$12.run(Cluster.java:2496)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at 
```

smiklosovic commented:
@bingqinzhou https://github.com/debezium/debezium-connector-cassandra/pull/31
